### PR TITLE
Possible fix for #1294 and #1370

### PR DESCRIPTION
**Pull request by seriva**
Possible fix for #1294 and #1370. Allows for named list to be merged:

Default:
```yaml
test:
  - name: group1
    value1: group1_1
    value2: group1_2
    value3: group1_3
  - name: group2
    value1: group2_1
    value2: group2_2
    value3: group2_3
  - name: group3
    value1: group3_1
    value2: group3_2
    value3: group3_3
  - name: group4
    value1: group4_1
    value2: group4_2
    value3: group4_3
```

Input:
```yaml
test:
  - name: group4
    value1: kaasbal1
    value2: kaasbal2
  - name: group5
    value1: group5_1
    value2: group5_2
    value3: group5_3
```

= output
```yaml
test:
  - name: group1
    value1: group1_1
    value2: group1_2
    value3: group1_3
  - name: group2
    value1: group2_1
    value2: group2_2
    value3: group2_3
  - name: group3
    value1: group3_1
    value2: group3_2
    value3: group3_3
  - name: group4
    value1: kaasbal1
    value2: kaasbal2
    value3: group4_3
  - name: group5
    value1: group5_1
    value2: group5_2
    value3: group5_3
```
Few notes
- Lists with no "name" fields in the list entries will use the default behavior of overriding the entire list as before. 
- Duplicate entries in named list are not allowed and will result in an error,  both default and input.
- Also added error when trying to merge fields of different types .

Affected defaults:
/workspaces/epiphany/core/src/epicli/data/aws/defaults/infrastructure/virtual-machine.yml
/workspaces/epiphany/core/src/epicli/data/azure/defaults/infrastructure/virtual-machine.yml
/workspaces/epiphany/core/src/epicli/data/common/defaults/configuration/applications.yml
/workspaces/epiphany/core/src/epicli/data/common/defaults/configuration/feature-mapping.yml
/workspaces/epiphany/core/src/epicli/data/common/defaults/configuration/grafana.yml
/workspaces/epiphany/core/src/epicli/data/common/defaults/configuration/haproxy.yml
/workspaces/epiphany/core/src/epicli/data/common/defaults/configuration/image-registry.yml
/workspaces/epiphany/core/src/epicli/data/common/defaults/configuration/kafka.yml
/workspaces/epiphany/core/src/epicli/data/common/defaults/configuration/postgresql.yml
/workspaces/epiphany/core/src/epicli/data/common/defaults/configuration/prometheus.yml
/workspaces/epiphany/core/src/epicli/data/common/defaults/configuration/vault.yml

### CODE REVIEW

**Review by erzetpe — DISMISSED** [review]

**Review by przemyslavic — DISMISSED** [review]

**Comment by przemyslavic** [comment]
I tested some different configurations for PostgreSQL, applications, feature-mapping. It looks like finally we have proper merging of dictionaries and lists.
What I especially like is enabling the applications in this simple way :)
```
---
kind: configuration/applications
title: "Kubernetes Applications Config"
provider: aws
name: default
specification:
  applications:
  - name: rabbitmq
    enabled: true
  - name: ignite-stateless
    enabled: true
  - name: auth-service
    enabled: true
  - name: pgpool
    enabled: false
  - name: pgbouncer
    enabled: false
  - name: istio
    enabled: false
```

File: core/src/epicli/cli/helpers/objdict_helpers.py [hunk 1]
```diff
@@ -32,6 +33,28 @@ def merge_objdict(to_merge, extend_by):
         if key in to_merge:
             if isinstance(to_merge[key], ObjDict):
                 merge_objdict(to_merge[key], val)
+            elif isinstance(to_merge[key], list):
```
Comment (to-bar) [inline core/src/epicli/cli/helpers/objdict_helpers.py]: This function is quite long, you can consider splitting to smaller ones to have self-documenting code (they would have names) or add a comment above the added part for people like me 😉

File: core/src/epicli/tests/helpers/test_objdict_helpers.py [hunk 2]
```diff
@@ -108,7 +109,7 @@ def test_dict_merge_updates_value_when_same_key_exists():
     assert base.field2 == 'val'
 
 
-def test_dict_merge_replaces_list_when_same_key_exists():
+def test_dict_merge_replaces_nonnamed_list_when_same_key_exists():
```
Comment (to-bar) [inline core/src/epicli/tests/helpers/test_objdict_helpers.py]:
```suggestion
def test_dict_merge_replaces_list_of_strings_when_same_key_exists():
```
Please consider adding additional test (e.g. `test_dict_merge_replaces_unnamed_list_when_same_key_exists`) for a list of hashes (without `name` key) instead of strings or keeping just single test but with hashes instead of strings.

File: core/src/epicli/tests/helpers/test_objdict_helpers.py [hunk 3]
```diff
@@ -117,6 +118,134 @@ def test_dict_merge_replaces_list_when_same_key_exists():
     assert base.list == ['replaced1', 'replaced2']
 
 
+def test_dict_merge_updates_named_list_when_same_name_key_exists():
```
Comment (to-bar) [inline core/src/epicli/tests/helpers/test_objdict_helpers.py:218]: :heart_eyes:

**Review by to-bar — CHANGES_REQUESTED** [review]

File: core/src/epicli/cli/helpers/objdict_helpers.py (hunk 1 above)
Comment (seriva) [reply core/src/epicli/cli/helpers/objdict_helpers.py]: - Split up function
- Documented it
- Added and restructured tests to reflect the new changes

Comment (seriva) [reply core/src/epicli/cli/helpers/objdict_helpers.py]: Broke it down, documented it a bit, added more tests to cover some edge case:)

**Review by erzetpe — DISMISSED** [review]

File: core/src/epicli/cli/helpers/objdict_helpers.py [hunk 4]
```diff
@@ -27,14 +28,53 @@ def objdict_to_dict(something):
...
+    return len(count) == len(l)
+
+
+def check_duplicate_in_named_list(l, key, value, type):
+        count = select_all(l, lambda x: x['name'] == value)
+        if len(count) > 1:
+            raise Exception(f'`name` field with value `"{value}"` occurs multiple times in list `"{key}"` in {type} definition.')
+
+
+#Note: We pas to_merge by reference using key to update/extend entries.
```
Comment (to-bar) [inline core/src/epicli/cli/helpers/objdict_helpers.py]:
```suggestion
# to_merge is passed by reference, item under key is updated, extended or replaced 
```

File: core/src/epicli/cli/helpers/objdict_helpers.py [hunk 5]
```diff
@@ -27,14 +28,53 @@ def objdict_to_dict(something):
...
+                for e_i in extend_by:
+                    check_duplicate_in_named_list(extend_by, key, e_i['name'], 'input')
+                    if m_i['name'] == e_i['name']:
+                        merge_objdict(m_i, e_i)
+                    else:
+                        count = select_all(to_merge[key], lambda x: x['name'] == e_i['name'])
+                        if len(count) == 0:
+                            to_merge[key].append(e_i)
+        else:
+            # No named list so just overwrite lists from with the source
```
Comment (to-bar) [inline core/src/epicli/cli/helpers/objdict_helpers.py]:
```suggestion
            # No named list so just replace item
```

File: core/src/epicli/cli/helpers/objdict_helpers.py [hunk 6]
```diff
@@ -27,14 +28,53 @@ def objdict_to_dict(something):
...
+
+
+#Note: We pas to_merge by reference using key to update/extend entries.
+def merge_list(to_merge, extend_by, key):
+        if is_named_list(to_merge[key]) and is_named_list(extend_by):
+            # Merge lists as named lists
+            for m_i in to_merge[key]:
+                check_duplicate_in_named_list(to_merge[key], key, m_i['name'], 'default')     
+                for e_i in extend_by:
+                    check_duplicate_in_named_list(extend_by, key, e_i['name'], 'input')
```
Comment (to-bar) [inline core/src/epicli/cli/helpers/objdict_helpers.py]: It looks like this function is called more times than necessary. How about moving loop into it and calling once for each list before loops? Is the bigger indentation of the first line of body by purpose?
```
def check_duplicate_in_named_list(list, key, type):
    for i in list:
        count = select_all(list, lambda x: x["name"] == i["name"])
        if len(count) > 1:
            raise Exception(
                f""""name" field with value "{i['name']}" occurs multiple times in list "{key}" in {type} definition."""
            )
```

Alternative version which can be faster:
```
def assert_unique_names_in_named_list(list, key, type):
    all_names = [x["name"] for x in list]

    for name in all_names:
        if all_names.count(name) > 1:
            raise Exception(
                f'"name" field with value "{name}" occurs multiple times in list "{key}" in {type} definition.'
            )
```

File: core/src/epicli/cli/helpers/objdict_helpers.py [hunk 7]
```diff
@@ -27,14 +28,53 @@ def objdict_to_dict(something):
     return _nested_dict_to_dict(something, src_class=dict, dst_class=dict)  # ObjDict is a subclass of dict
 
 
+def is_named_list(l):
+    count = select_all(l, lambda x: hasattr(x, 'name'))
+    return len(count) == len(l)
```
Comment (to-bar) [inline core/src/epicli/cli/helpers/objdict_helpers.py]: One-liner alternative:
```suggestion
    return all(hasattr(x, 'name') for x in l)
```

File: core/src/epicli/cli/helpers/objdict_helpers.py [hunk 8]
```diff
@@ -27,14 +28,53 @@ def objdict_to_dict(something):
...
-            if isinstance(to_merge[key], ObjDict):
+            if isinstance(to_merge[key], ObjDict) and isinstance(val, ObjDict):
+                # Dealing with 2 ObjDicts so recursively merge
                 merge_objdict(to_merge[key], val)
-            else:
+            elif isinstance(to_merge[key], list) and isinstance(val, list):
+                # Dealing with 2 lists
+                merge_list(to_merge, val, key)
+            elif type(to_merge[key]) == type(val):
+                # Dealing with 2 base fields (integer, boolean, string ..etc) so overwrite from source to defaults
```
Comment (to-bar) [inline core/src/epicli/cli/helpers/objdict_helpers.py]:
```suggestion
                # Dealing with 2 basic types (integer, boolean, string, etc.) so replace
```

File: core/src/epicli/cli/helpers/objdict_helpers.py [hunk 9]
```diff
@@ -27,14 +28,53 @@ def objdict_to_dict(something):
...
                 to_merge[key] = val
+            else:
+                # If we come here we are dealing with 2 different types we cannot merge so throw exception.
+                raise Exception(f'Types of key `"{key}"` are different: {type(to_merge[key])}, {type(val)}. Unable to merge.')
```
Comment (to-bar) [inline core/src/epicli/cli/helpers/objdict_helpers.py]:
```suggestion
                raise Exception(f'Types of key "{key}" are different: {type(to_merge[key])}, {type(val)}. Unable to merge.')
```

File: core/src/epicli/tests/helpers/test_objdict_helpers.py [hunk 10]
```diff
@@ -90,6 +91,51 @@ def test_objdict_to_dict():
...
+
+    assert is_named_list(l['list'])
+
+
+def test_list_is_not_named_list():
+    l = dict_to_objdict({
+        'list': [
+            'entry1',
+            'entry2'           
+        ]
```
Comment (to-bar) [inline core/src/epicli/tests/helpers/test_objdict_helpers.py]:
```suggestion
        'list': [
            {
                'name': 'name1',
                'value': 'value1'
            },
            {
                'value': 'value2'
            }             
        ]
```

File: core/src/epicli/tests/helpers/test_objdict_helpers.py [hunk 11]
```diff
@@ -90,6 +91,51 @@ def test_objdict_to_dict():
...
+    })
+
+    assert is_named_list(l['list']) == False
+
+
+def test_check_assert_in_duplicate_in_named_list():
```
Comment (to-bar) [inline core/src/epicli/tests/helpers/test_objdict_helpers.py]:
```suggestion
def test_assert_unique_names_in_named_list():
```

**Review by to-bar — CHANGES_REQUESTED** [review]

File: core/src/epicli/cli/helpers/objdict_helpers.py (hunk 6 above)
Comment (seriva) [reply core/src/epicli/cli/helpers/objdict_helpers.py]: Good point, will check it out!

File: core/src/epicli/cli/helpers/objdict_helpers.py (hunk 7 above)
Comment (seriva) [reply core/src/epicli/cli/helpers/objdict_helpers.py]: Nice! will test and apply!

File: core/src/epicli/cli/helpers/objdict_helpers.py [hunk 12]
```diff
@@ -27,14 +28,52 @@ def objdict_to_dict(something):
     return _nested_dict_to_dict(something, src_class=dict, dst_class=dict)  # ObjDict is a subclass of dict
 
 
+def is_named_list(l):
+    return all(hasattr(x, 'name') for x in l)
```
Comment (to-bar) [inline core/src/epicli/cli/helpers/objdict_helpers.py]: We can also handle non-iterable objects (such as integers) to avoid errors like `TypeError: 'int' object is not iterable`. The same affects the previous version of this function.

```suggestion
    return isinstance(l, Iterable) and all(hasattr(x, 'name') for x in l)
```

`Iterable` requires importing: `from collections.abc import Iterable`

File: core/src/epicli/cli/helpers/objdict_helpers.py (hunk 6 above)
Comment (seriva) [reply core/src/epicli/cli/helpers/objdict_helpers.py]: Optimized todo the merge in 2 passes instead of nested loops, much more efficient:)  Also makes the code a lot easier to follow imo.

**Review by erzetpe — DISMISSED** [review]

**Review by erzetpe — DISMISSED** [review]

File: core/src/epicli/tests/helpers/test_objdict_helpers.py [hunk 13]
```diff
@@ -90,6 +91,56 @@ def test_objdict_to_dict():
...
+                'value': 'value1'
+            },
+            {
+                'name': 'name1',
+                'value': 'value2'
+            }             
+        ]
+    })
+    with pytest.raises(Exception) as e_info:
+        assert_duplicate_names_in_named_list(l['list'], 'list', input)
```
Comment (to-bar) [inline core/src/epicli/tests/helpers/test_objdict_helpers.py]:
```suggestion
        assert_duplicate_names_in_named_list(l['list'], 'list', 'input')
```

File: core/src/epicli/tests/helpers/test_objdict_helpers.py [hunk 14]
```diff
@@ -151,3 +204,140 @@ def test_dict_merge_add_field_to_nested_object():
...
+                'name': 'test2',
+                'value': 'base2'
+            },
+            {
+                'name': 'test3',
+                'value': 'extend3'
+            }]
+
+
+def test_dict_merge_asserts_on_base_named_list_with_duplicate_key():
```
Comment (to-bar) [inline core/src/epicli/tests/helpers/test_objdict_helpers.py]:
```suggestion
def test_dict_merge_asserts_on_base_named_list_with_duplicate_name():
```

File: core/src/epicli/tests/helpers/test_objdict_helpers.py [hunk 15]
```diff
@@ -151,3 +204,140 @@ def test_dict_merge_add_field_to_nested_object():
...
+                'name': 'test3',
+                'value': 'extend3'
+            }            
+        ]
+    })
+    with pytest.raises(DuplicatesInNamedListException):
+        merge_objdict(base, extend_by)
+
+
+def test_dict_merge_asserts_on_extend_named_list_with_duplicate_key():
```
Comment (to-bar) [inline core/src/epicli/tests/helpers/test_objdict_helpers.py]:
```suggestion
def test_dict_merge_asserts_on_extend_named_list_with_duplicate_name():
```

File: core/src/epicli/cli/helpers/objdict_helpers.py [hunk 16]
```diff
@@ -27,14 +37,57 @@ def objdict_to_dict(something):
...
+    all_names = [x["name"] for x in list] 
+    for name in all_names: 
+        if all_names.count(name) > 1: 
+            raise DuplicatesInNamedListException( f'"name" field with value "{name}" occurs multiple times in list "{key}" in {type} definition.' )
+
+
+# to_merge is passed by reference, item under key is updated, extended or replaced 
+def merge_list(to_merge, extend_by, key):
+    if is_named_list(to_merge[key]) and is_named_list(extend_by):
+        # check for duplicates items in to_merge and extend_by
```
Comment (to-bar) [inline core/src/epicli/cli/helpers/objdict_helpers.py]:
```suggestion
        # ensure all items have unique names in to_merge and extend_by
```

File: core/src/epicli/cli/epicli.py [hunk 17]
```diff
@@ -418,9 +420,16 @@ def ensure_vault_password_is_cleaned():
         os.remove(Config().vault_password_location)
 
 
+def total_run_time():
```
Comment (to-bar) [inline core/src/epicli/cli/epicli.py]:
```suggestion
def log_total_run_time():
```

File: core/src/epicli/cli/epicli.py [hunk 18]
```diff
@@ -418,9 +420,16 @@ def ensure_vault_password_is_cleaned():
...
+    logger = Log('run_time')
+    passed_time = format_time(time.time()-start_time)
+    logger.info(f'Total run time: {passed_time}')
+
+
 def exit_handler():
     ensure_vault_password_is_cleaned()
-
+    total_run_time()
```
Comment (to-bar) [inline core/src/epicli/cli/epicli.py]:
```suggestion
    log_total_run_time()
```

**Review by to-bar — DISMISSED** [review]

**Review by erzetpe — DISMISSED** [review]

**Review by erzetpe — DISMISSED** [review]

**Review by to-bar — DISMISSED** [review]

**Comment by seriva** [comment]
Lets have @przemyslavic give this one more test spin before we merge it.

/azp run

**Review by erzetpe — DISMISSED** [review]

**Review by to-bar — DISMISSED** [review]

File: core/src/epicli/cli/helpers/time_helpers.py [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
+    secs = seconds%60
+    minutes = floor((seconds/60)%60)
+    hours = floor(seconds/(60*60)%24)
+    if hours != 0:
+        return "%dh %dm %.2fs" % (hours, minutes, secs)
+    elif hours == 0 and minutes != 0:
+        return "%dm %.2fs" % (minutes, secs)
+    else:
+        return "%.2fs" % (secs)
```
Comment (to-bar) [inline core/src/epicli/cli/helpers/time_helpers.py]:
```suggestion
def format_time(secs):
    """Formats seconds to a time string with the format: "Xh Xm X.XXs", "Xm X.XXs", "X.XXs"

    Parameters:
    secs (float): Time in seconds

    Returns:
    string: Formatted time string
   """

    seconds = secs%60
    minutes = floor((secs/60)%60)
    hours = floor(secs/(60*60)%24)
    if hours != 0:
        return "%dh %dm %.2fs" % (hours, minutes, seconds)
    elif hours == 0 and minutes != 0:
        return "%dm %.2fs" % (minutes, seconds)
    else:
        return "%.2fs" % (seconds)
```

**Review by to-bar — APPROVED** [review]

**Review by erzetpe — APPROVED** [review]
